home: Handle invalid language before completing the request.

sahil839 opened the pull request:
This PR adds code to handle default languages set to values which we do not support
due to less than 5% of strings being translated for that language.

If realm's default langauge is invalid then we set it to "en". And if user's default langauge
is invalid, we set it to the realm default.

<!-- Describe your pull request here.-->

<!-- Issue link, or clear description.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/lib/home.py:
```diff
@@ -104,6 +107,10 @@ def build_page_params_for_home_page_load(
 
+    available_language_codes = get_available_language_codes()
+    if realm.default_language not in available_language_codes:
+        do_set_realm_property(realm, "default_language", "en", acting_user=None)
```
Let's add a comment here:

> If we no longer have a translation for the organization's language, correct the organization to have one; this block is a workaround for there not being a good time to run a migration to audit this after a new server version is deployed.

And have it log at warning level when this happens, maybe:

> Realm {string_id} has an unsupported default language ({language_code}), reverting to English.

timabbott commented on zerver/lib/home.py:
```diff
@@ -124,6 +131,12 @@ def build_page_params_for_home_page_load(
+                user_profile, "default_language", realm.default_language, acting_user=None
+            )
+            default_language = user_profile.default_language
+            state_data["user_settings"]["default_language"] = user_profile.default_language
```
I think this block deserves a comment explaining the purpose/reasoning here.

BTW should this logic be in `fetch_initial_state_data` instead so it'll apply for mobile clients registering too? Ideally with some care to not fetch it for random bots registering just for messages as an event type.

sahil839 commented on zerver/tests/test_home.py:
```diff
@@ -1401,6 +1402,35 @@ def test_realm_push_notifications_enabled_end_timestamp(self) -> None:
+        cordelia = self.example_user("cordelia")
+        hamlet = self.example_user("hamlet")
+        do_change_user_setting(cordelia, "default_language", "gl", acting_user=None)
+        do_change_user_setting(hamlet, "default_language", "no", acting_user=None)
```
I think we can mock `get_language_list` which just returns the "name_map" object value from `language_name_map.json`.

```py
@lru_cache(None)
def get_language_list() -> list[dict[str, Any]]:
    path = os.path.join(settings.DEPLOY_ROOT, "locale", "language_name_map.json")
    with open(path, "rb") as reader:
        languages = orjson.loads(reader.read())
        return languages["name_map"]
```

sahil839 commented:
Updated the PR.
- Move code to update language in `fetch_initial_state_data`.
- Added comments.
- Using mocked language list for tests. Also updated other tests for updating the language to use mocked list.
